localstack/services/routing.py

```diff
@@ -0,0 +1,175 @@
+import dataclasses
+import json
+from typing import Any, Dict, List, Optional, Tuple, Union
+from urllib.parse import urlparse
+
+from localstack.services.generic_proxy import ProxyListener
+from localstack.utils.common import to_str
+
+
+@dataclasses.dataclass
+class Request:
+    method: str
+    path: str
+    data: Union[str, bytes]
+    headers: Dict
+
+    @property
+    def host(self) -> Optional[str]:
+        return self.headers.get("Host")
+
+    def json(self):
+        return json.loads(to_str(self.data or "{}"))
+
+
+@dataclasses.dataclass
+class RoutingRule:
+    uri_template: str
+    is_pattern: bool = False
+    method: str = None
+    match_host: bool = False
+
+    def url(self):
+        return urlparse(self.uri_template)
+
+    def matches(self, request: Request) -> bool:
+        if self.method and self.method != request.method:
+            return False
+
+        if self.is_pattern:
+            raise NotImplementedError
```

Comment:
these will be regex-like patterns or what's the plan for this feature?

Reply:
or will this be something like `/cloudformation/resources/<resource>/state` (just random example) where the resource would be mapped to a parameter on the handler function?

Reply:
the latter :-)